piv/piv.go

```diff
@@ -476,19 +517,48 @@ func (yk *YubiKey) SetManagementKey(oldKey, newKey [24]byte) error {
 
 // ykSetManagementKey updates the management key to a new key. This requires
 // authenticating with the existing management key.
-func ykSetManagementKey(tx *scTx, key [24]byte, touch bool) error {
+func ykSetManagementKey(tx *scTx, key []byte, touch bool) error {
+	var managementKeyType byte
+	switch len(key) {
+	case 16:
+		managementKeyType = algAES128
+	case 24:
+		// could also be 3DES key, but AES192 is preferred so try that first
+		managementKeyType = algAES192
+	case 32:
+		managementKeyType = algAES256
+	default:
+		return fmt.Errorf("invalid new management key length: %d", len(key))
+	}
 	cmd := apdu{
 		instruction: insSetMGMKey,
 		param1:      0xff,
 		param2:      0xff,
 		data: append([]byte{
-			alg3DES, keyCardManagement, 24,
+			managementKeyType, keyCardManagement, byte(len(key)),
 		}, key[:]...),
 	}
 	if touch {
 		cmd.param2 = 0xfe
 	}
-	if _, err := tx.Transmit(cmd); err != nil {
+	_, err := tx.Transmit(cmd)
+	if err != nil && len(key) == 24 {
+		// if setting AES192 managment key failed, fallback to 3DES
```

Comment:
I think this is too "magic". If you fail to use AES for a management key then you would want to know about it with an error. Not just fall back to the deprecated 3DES.

IMO the version check algorithm described in https://github.com/go-piv/piv-go/issues/146#issuecomment-2176449257 is a better approach.

Reply:
If you fail and have to fall back, it's because the token doesn't support AES. In that case 3DES isn't deprecated - it's the only key type that is supported. I'm not opposed to changing this to be guarded by a version check instead, I chose to do it this way to not be as dependent on yubikey specific versioning as it appears people use this for more than just yubico brand tokens.